File: examples/python/shift-ops/ShiftOps.runar.py
"""ShiftOps -- Exercises bitshift operators ``<<`` and ``>>`` on Bigint values."""

from runar import SmartContract, Bigint, public, assert_


class ShiftOps(SmartContract):
    """Exercises bitshift operators on Bigint values."""

    a: Bigint

    def __init__(self, a: Bigint):
        super().__init__(a)
        self.a = a

    @public
    def test_shift(self):
        """Apply left shift and right shift, then sanity-check the results."""
        left = self.a << 3
        right = self.a >> 2
        assert_(left >= 0 or left < 0)
        assert_(right >= 0 or right < 0)

File: examples/zig/shift-ops/ShiftOps.runar.zig
const runar = @import("runar");

pub const ShiftOps = struct {
    pub const Contract = runar.SmartContract;

    a: i64,

    pub fn init(a: i64) ShiftOps {
        return .{ .a = a };
    }

    pub fn testShift(self: *const ShiftOps) void {
        const left = self.a << 3;
        const right = self.a >> 2;
        runar.assert(left >= 0 or left < 0);
        runar.assert(right >= 0 or right < 0);
    }
};
